pkg/activator/throttler.go

```diff
@@ -0,0 +1,151 @@
+package activator
+
+import (
+	"github.com/knative/serving/pkg/queue"
+	"go.uber.org/zap"
+	"sync"
+	"github.com/knative/serving/pkg/client/informers/externalversions/serving/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+	v1alpha12 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"errors"
+	"strings"
+	corev1 "k8s.io/api/core/v1"
+	"fmt"
+)
+
+// NewThrottler creates a new Throttler
+func NewThrottler(queueDepth, maxConcurrency int32, revisionInformer v1alpha1.RevisionInformer, logger *zap.SugaredLogger) *Throttler {
+	breakers := make(map[RevisionID]queue.Breaker)
+	return &Throttler{breakers: breakers, queueDepth: queueDepth, maxConcurrency: maxConcurrency, revisionInformer: revisionInformer, logger: logger}
+}
+
+// Throttler allows to update the max concurrency of registered Breakers
+type Throttler struct {
+	breakers         map[RevisionID]queue.Breaker
+	revisionInformer v1alpha1.RevisionInformer
+	queueDepth       int32
+	maxConcurrency   int32
+	logger           *zap.SugaredLogger
+	mux              sync.Mutex
+}
+
+// Add potentially creates a new breaker for the revision
+// it returns true if the breaker already exists, false otherwise
+func (t *Throttler) Add(rev RevisionID) bool {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	if _, ok := t.breakers[rev]; !ok {
+		breaker := queue.NewBreaker(t.queueDepth, t.maxConcurrency, 0)
+		t.breakers[rev] = *breaker
+		return ok
+	}
+	return true
+}
+
+// Exists returns true if a Breaker already exists for the given revision
+func (t *Throttler) Exists(rev RevisionID) bool {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	_, ok := t.breakers[rev]
+	return ok
+}
+
+// Removes the breaker from the bookkeeping
+func (t *Throttler) Remove(rev RevisionID) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	delete(t.breakers, rev)
+}
+
+// UpdateCapacity update the max concurrency of the Breaker corresponding to the revision
+func (t *Throttler) UpdateCapacity(rev RevisionID, size int32) error {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	var containerConcurrency int32
+	// Knative Revision and k8s service have different names
+	baseName := strings.Split(rev.Name, "-service")[0]
+	revision, err := t.getRevision(rev.Namespace, baseName, t.revisionInformer)
+	if err != nil {
+		return errors.New(fmt.Sprintf("No revision was found for: %s", rev))
+	} else {
+		containerConcurrency = int32(revision.Spec.ContainerConcurrency)
+		// Equals to the pending queue size as defined in `cmd/queue/main.go`
+		if containerConcurrency < 10 {
+			containerConcurrency = 10
+		}
```

Comment:
The pending queue size is small on the containers themselves for a reason. As the pending queue size for the breakers are hard-wired to be 1000 here (which is fine for starters), I think we can just drop this guard.

Reply:
Which concurrency would you assign then? The pending queue is at least 10, if we give more then that, there is a chance of requests circling in the retry loop.

Reply:
Isn't it that we will break the ones who define their container concurrency to be > 10? (sorry, maybe I'm just floundering here)

Reply:
In the queue, this defines the `queueDepth` (that is: the pending queue). The breaker itself breaks on the "normal" containerConcurrency.

Only the queueDepth is guarded by this `if` block. It shouldn't be needed here.